### ISSUE TITLE
Inconsistent output of `flatpak ps`

### ISSUE THREAD

**Issue by fbruetting** [flatpak/flatpak#3246]
* Linux distribution and version: FSB 31
* Flatpak version: v1.4.3

## Description of the problem

Somehow, the output of `flatpak ps` differs between the used terminal applications and the point in time. On my freshly booted system, flatpaked Tilix shows nothing, whereas the Tilix-RPM shows me seven consistent applications:

```
❯ flatpak ps
Instance   PID   Application                Runtime
3808991453 28787 org.gnome.Calendar         org.gnome.Platform
186606205  28066 org.mozilla.Firefox        org.fedoraproject.Platform
2387262295 19113 com.spotify.Client         org.freedesktop.Platform
1410901933 13246 org.gnome.Lollypop         org.gnome.Platform
2531377316 8994  org.gnome.Evolution        org.gnome.Platform
345940244  2068  com.gitlab.bitseater.meteo org.gnome.Platform
2281028135 2118  im.riot.Riot               org.freedesktop.Platform
```

However, sometimes when I repeatedly run that command, I see gedit instead of Calendar, but the amount of displayed programs doesn’t change at all.

Even more strange is running that command in Gnome-Terminal, as there it starts with 12 programs:

```
❯ flatpak ps
Instance   PID   Application                Runtime
1523127705 28792 org.gnome.clocks           org.gnome.Platform
384848805  28785 de.haeckerfelix.gradio     org.gnome.Platform
3808991453 28787 org.gnome.Calendar         org.gnome.Platform
1141320892 28786 org.gnome.Boxes            org.gnome.Platform
842683723  28788 org.gnome.Contacts         org.gnome.Platform
186606205  28066 org.mozilla.Firefox        org.fedoraproject.Platform
342201308  27913 com.gexperts.Tilix         org.fedoraproject.Platform
2387262295 19113 com.spotify.Client         org.freedesktop.Platform
1410901933 13246 org.gnome.Lollypop         org.gnome.Platform
2531377316 8994  org.gnome.Evolution        org.gnome.Platform
345940244  2068  com.gitlab.bitseater.meteo org.gnome.Platform
2281028135 2118  im.riot.Riot               org.freedesktop.Platform
```

BUT: The more often I enter the command, the fewer applications get displayed until it went down to seven, then it stays at seven:

```
❯ flatpak ps | wc -l
13

~ 
❯ flatpak ps | wc -l
12

~ 
❯ flatpak ps | wc -l
12

~ 
❯ flatpak ps | wc -l
12

~ took 3s 
❯ flatpak ps | wc -l
12

~ 
❯ flatpak ps | wc -l
12

~ 
❯ flatpak ps | wc -l
12

~ 
❯ flatpak ps | wc -l
9

~ 
❯ flatpak ps | wc -l
9

~ 
❯ flatpak ps | wc -l
8
```

I can run that command on Tilix-RPM and on Gnome-Terminal and swap between both programs, but the output would not change. So when I open both and run it on Tilix and immediately afterwards on GT, Tilix displays 7 programs and GT displays 12 (and then goes down).

Opening a new pane on GT changes nothing, output stays at seven programs – I have to restart GT for the count of running Flatpaks to go back to 12. How can this be?

**Comment by matthiasclasen** [comment]
flatpak ps is essentially a fancy way to do

```
ls /run/user/1000/.flatpak
```

Can you compare the output with that ?

Running flatpak ps with --verbose might provide some hints as to what is going on.

**Comment by fbruetting** [comment]
That’s really weird: The listing of `ls /run/user/1000/.flatpak` also varies with time… 😳️ It seems to me that something triggers the generation of the full listing there and then over time the list gets shortened by some amount.

**Comment by matthiasclasen** [comment]
We do leave directories behind in /run/usr/1000/.flatpak because there is no good hook for us to notice when a flatpak exits. flatpak ps checks which directories still have an active running flatpak associated with them and clean stale ones up.

**Comment by fbruetting** [comment]
But why does that get reset to 13 and then those folders get removed down to 8? Does that mean that running flatpaks get falsely detected as not-running and their respective folders thus removed?

**Comment by matthiasclasen** [comment]
I don't think that is the case. I have no idea what you are seeing, though.

**Comment by fbruetting** [comment]
I can also use Tilix-RPM or GNOME Terminal and see the same.

**Comment by matthiasclasen** [comment]
If you run `flatpak ps --verbose`, it will tell you when it finds and cleans up unused directories in /run/user/UID/.flatpap:

```
$ flatpak ps --verbose 
F: Cleaning up unused container id 2062338624
3148562640	798026	org.gnome.Polari	org.gnome.Platfor

```

**Comment by fbruetting** [comment]
Yes, here it is. I first opened GNOME Terminal and typed it a few times (1), then opened Tilix (RPM) and typed it it once (2) – where you can see that some packages were cleaned somehow, and then also GNOME Terminal refects that change.

![Bildschirmfoto von 2019-12-09 01-22-09](https://user-images.githubusercontent.com/10520767/70399291-fdf31d00-1a22-11ea-9f1f-db1d530f3b8a.png)

**Comment by matthiasclasen** [comment]
We are looking at the mtime of the .ref file inside these directories, and if it shows up as older than 3 seconds, we clean up the unused directory. The only idea I have is that somehow the terminal on the right (in a sandbox, right ?) fails to get the proper timestamps ? Maybe due to some fuse / namespacing shenanigans ?

**Comment by alexlarsson** [comment]
Its pretty weird that it acts differently in the other terminal. Can you think of any other differences between the two than the GUI app that runs the terminal. Like, different namespaces, different shells, env vars, etc?

**Comment by fbruetting** [comment]
> the terminal on the right (in a sandbox, right ?)

No, Tilix is installed as an RPM.

What I found out is, that Tilix and GNOME Terminal behave indentically. These zombie processes get created as soon as I open GNOME Shell’s overview via pressing `Super` and typing in some letters. It doesn’t matter if I open an application after typing in some letters or if I cancel the overview by pressing `Super` or `Esc`, bit I have to type in some letters. Without typing in some letters, nothing happens.

**Comment by chrisawi** [comment]
> What I found out is, that these “zombie” processes get created as soon as I open GNOME Shell’s overview via pressing `Super` AND then at every time I type in one letter. I created a bash loop showing me the `flatpak ps -v` output repeatedly. Then I opened the overview, typed in one letter and waited until the “zombie” processes disappeared. Then I typed in a second letter and suddenly all those wild “zombie” processes appeared again.

This all sounds totally normal. The search providers are dbus-activated and stick around for 10s before exiting.

I'm not sure there's a bug here.
